File: packages/engine/src/cpus/6502/instructions/adc.test.js
import { test, expect } from 'vitest'
import { BufferMapper } from '../../../buffer_mapper'
import { CPU } from '..'

test('adc imm', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.acc = 2
  mmu.data = [0x69, 3] // ADC #$03
  const g = cpu.run()
  for (let i = 0; i < 2; i++) g.next()
  expect(cpu.acc).toBe(5)
})

test('adc abs', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.acc = 2
  mmu.data = [0x6d, 0x03, 0x0, 1] // ADC $03
  const g = cpu.run()
  for (let i = 0; i < 4; i++) g.next()
  expect(cpu.acc).toBe(3)
})

test('adc abx', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.acc = 2
  cpu.x = 2
  mmu.data = [0x7d, 0x01, 0x0, 1] // ADC X+$01
  const g = cpu.run()
  for (let i = 0; i < 4; i++) g.next()
  expect(cpu.acc).toBe(3)
})

test('adc abx + page crossing', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.acc = 2
  cpu.x = 1
  mmu.data = [0x7d, 0xff, 0x0] // ADC X+$FF=$100
  mmu.data[0x100] = 3
  const g = cpu.run()
  for (let i = 0; i < 5; i++) g.next()
  expect(cpu.acc).toBe(5)
})

test('adc aby', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.acc = 2
  cpu.y = 2
  mmu.data = [0x79, 0x01, 0x0, 1] // ADC Y+$01
  const g = cpu.run()
  for (let i = 0; i < 4; i++) g.next()
  expect(cpu.acc).toBe(3)
})

test('adc zpg', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.acc = 2
  mmu.data = [0x65, 0x02, 1] // ADC $02
  const g = cpu.run()
  for (let i = 0; i < 3; i++) g.next()
  expect(cpu.acc).toBe(3)
})

test('adc zpx', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.acc = 2
  cpu.x = 1
  mmu.data = [0x75, 0x02, 0, 1] // ADC X+$02
  const g = cpu.run()
  for (let i = 0; i < 4; i++) g.next()
  expect(cpu.acc).toBe(3)
})

test('adc idx', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.acc = 5
  cpu.x = 1
  mmu.data = [0x61, 0xff] // ADC [[X+#$FF]]=$FF02
  mmu.data[0x4] = 2
  mmu.data[0x5] = 0xff
  mmu.data[0xff02] = 3
  const g = cpu.run()
  for (let i = 0; i < 6; i++) g.next()
  expect(cpu.acc).toBe(5)
})

test('adc idy', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.acc = 2
  cpu.y = 1
  mmu.data = [0x71, 0xfe] // ADC Y+[[#$FF]=$FF02]
  mmu.data[0xfe] = 2
  mmu.data[0xff] = 0xff
  mmu.data[0xff03] = 3
  const g = cpu.run()
  for (let i = 0; i < 5; i++) g.next()
  expect(cpu.acc).toBe(5)
})

test('adc idy + page crossing', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.acc = 2
  cpu.y = 1
  mmu.data = [0x71, 0xfe] // ADC Y+[[#$FF]=$FF02]
  mmu.data[0xfe] = 0xff
  mmu.data[0xff] = 0xef
  mmu.data[0xf000] = 3
  const g = cpu.run()
  for (let i = 0; i < 6; i++) g.next()
  expect(cpu.acc).toBe(5)
})

File: packages/engine/src/cores/famibob/cpu/instructions/adc.test.zig
const testing = @import("../testing/setup.zig");

test "ADC imm" {
    var cpu = testing.createCPU();
    cpu.acc = 2;
    cpu.write(0, 0x69);
    cpu.write(1, 3);
    for (0..2) |_| {
        cpu.process();
    }
    try testing.expectEqual(5, cpu.acc);
}

test "ADC abs" {
    var cpu = testing.createCPU();
    cpu.acc = 2;
    cpu.write(0, 0x6d);
    cpu.write(1, 3);
    cpu.write(2, 0);
    cpu.write(3, 1);
    for (0..4) |_| {
        cpu.process();
    }
    try testing.expectEqual(3, cpu.acc);
}

test "ADC abx" {
    var cpu = testing.createCPU();
    cpu.acc = 2;
    cpu.x = 2;
    cpu.write(0, 0x7d);
    cpu.write(1, 1);
    cpu.write(2, 0);
    cpu.write(3, 1);
    for (0..4) |_| {
        cpu.process();
    }
    try testing.expectEqual(3, cpu.acc);
}

test "ADC abx + page crossing" {
    var cpu = testing.createCPU();
    cpu.acc = 2;
    cpu.x = 1;
    cpu.write(0, 0x7d);
    cpu.write(1, 0xff);
    cpu.write(2, 0);
    cpu.write(0x100, 3);
    for (0..5) |_| {
        cpu.process();
    }
    try testing.expectEqual(5, cpu.acc);
}

test "ADC aby" {
    var cpu = testing.createCPU();
    cpu.acc = 2;
    cpu.y = 2;
    cpu.write(0, 0x79);
    cpu.write(1, 1);
    cpu.write(2, 0);
    cpu.write(3, 1);
    for (0..4) |_| {
        cpu.process();
    }
    try testing.expectEqual(3, cpu.acc);
}

test "ADC zpg" {
    var cpu = testing.createCPU();
    cpu.acc = 2;
    cpu.write(0, 0x65);
    cpu.write(1, 2);
    cpu.write(2, 1);
    for (0..3) |_| {
        cpu.process();
    }
    try testing.expectEqual(3, cpu.acc);
}

test "ADC zpx" {
    var cpu = testing.createCPU();
    cpu.acc = 2;
    cpu.x = 1;
    cpu.write(0, 0x75);
    cpu.write(1, 2);
    cpu.write(2, 0);
    cpu.write(3, 1);
    for (0..4) |_| {
        cpu.process();
    }
    try testing.expectEqual(3, cpu.acc);
}

test "ADC idx" {
    var cpu = testing.createCPU();
    cpu.acc = 5;
    cpu.x = 1;
    cpu.write(0, 0x61);
    cpu.write(1, 0xff);
    cpu.write(4, 2);
    cpu.write(5, 0xff);
    cpu.write(0xff02, 3);
    for (0..6) |_| {
        cpu.process();
    }
    try testing.expectEqual(5, cpu.acc);
}

test "ADC idy" {
    var cpu = testing.createCPU();
    cpu.acc = 2;
    cpu.y = 1;
    cpu.write(0, 0x71);
    cpu.write(1, 0xfe);
    cpu.write(0xfe, 2);
    cpu.write(0xff, 0xff);
    cpu.write(0xff03, 3);
    for (0..5) |_| {
        cpu.process();
    }
    try testing.expectEqual(5, cpu.acc);
}

test "ADC idy + page crossing" {
    var cpu = testing.createCPU();
    cpu.acc = 2;
    cpu.y = 1;
    cpu.write(0, 0x71);
    cpu.write(1, 0xfe);
    cpu.write(0xfe, 0xff);
    cpu.write(0xff, 0xef);
    cpu.write(0xf000, 3);
    for (0..6) |_| {
        cpu.process();
    }
    try testing.expectEqual(5, cpu.acc);
}
